google/cloud/internal/oauth2_refreshing_credentials_wrapper_test.cc

```diff
@@ -0,0 +1,163 @@
+#include "google/cloud/testing_util/status_matchers.h"
+#include <gmock/gmock.h>
+
+namespace google {
+namespace cloud {
+namespace oauth2_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace {
+
+using ::google::cloud::testing_util::IsOk;
+using ::testing::Eq;
+using ::testing::Not;
+
+TEST(RefreshingCredentialsWrapper, IsExpired) {
+  std::pair<std::string, std::string> const auth_token =
+      std::make_pair("Authorization", "foo");
+  RefreshingCredentialsWrapper w;
+  auto refresh_fn = [&]() {
+    RefreshingCredentialsWrapper::TemporaryToken token;
+    token.token = auth_token;
+    token.expiration_time =
+        std::chrono::system_clock::now() - std::chrono::seconds(1000);
+    return token;
+  };
+  auto token =
+      w.AuthorizationHeader(std::chrono::system_clock::now(), refresh_fn);
+  EXPECT_TRUE(w.IsExpired(std::chrono::system_clock::now()));
+}
+
+TEST(RefreshingCredentialsWrapper, IsNotExpired) {
+  std::pair<std::string, std::string> const auth_token =
+      std::make_pair("Authorization", "foo");
+  RefreshingCredentialsWrapper w;
+  auto refresh_fn = [&]() {
+    RefreshingCredentialsWrapper::TemporaryToken token;
+    token.token = auth_token;
+    token.expiration_time =
+        std::chrono::system_clock::now() + std::chrono::minutes(60);
+    return token;
+  };
+  auto token =
+      w.AuthorizationHeader(std::chrono::system_clock::now(), refresh_fn);
+
+  EXPECT_FALSE(w.IsExpired(std::chrono::system_clock::now()));
+}
+
+TEST(RefreshingCredentialsWrapper, IsValid) {
+  std::pair<std::string, std::string> const auth_token =
+      std::make_pair("Authorization", "foo");
+  RefreshingCredentialsWrapper w;
+  auto refresh_fn = [&]() {
+    RefreshingCredentialsWrapper::TemporaryToken token;
+    token.token = auth_token;
+    token.expiration_time =
+        std::chrono::system_clock::now() + std::chrono::minutes(60);
+    return token;
+  };
+  auto token =
+      w.AuthorizationHeader(std::chrono::system_clock::now(), refresh_fn);
+  EXPECT_TRUE(w.IsValid(std::chrono::system_clock::now()));
+}
+
+TEST(RefreshingCredentialsWrapper, IsNotValid) {
+  std::pair<std::string, std::string> const auth_token =
+      std::make_pair("Authorization", "foo");
+  RefreshingCredentialsWrapper w;
+  EXPECT_FALSE(w.IsValid(std::chrono::system_clock::now()));
+}
+
+TEST(RefreshingCredentialsWrapper, RefreshTokenSuccess) {
+  std::pair<std::string, std::string> const auth_token =
+      std::make_pair("Authorization", "foo");
+  RefreshingCredentialsWrapper w;
+  auto refresh_fn = [&]() {
+    RefreshingCredentialsWrapper::TemporaryToken token;
+    token.token = auth_token;
+    token.expiration_time =
+        std::chrono::system_clock::now() + std::chrono::seconds(60);
+    return token;
+  };
+  auto token =
+      w.AuthorizationHeader(std::chrono::system_clock::now(), refresh_fn);
+  ASSERT_THAT(token, IsOk());
+  EXPECT_THAT(*token, Eq(auth_token));
+}
```

Comment:
Consider a second call to `w.AuthorizationHeader()` to test the case where the token is valid and does not need a refresh.